windows/src/engine/keyman32/keyman32.cpp

```diff
@@ -563,64 +563,55 @@ extern "C" BOOL  _declspec(dllexport) WINAPI Keyman_ForceKeyboard(PCSTR FileName
 
-    if (err_status != KM_KBP_STATUS_OK) {
-      SendDebugMessageFormat(
-          0, sdmGlobal, 0, "Keyman_ForceKeyboard Core: km_kbp_state_create failed with error status [%d]", err_status);
-      // Dispose of the keyboard to leave us in a consitent state
-      ReleaseKeyboardMemoryCore(&_td->lpActiveKeyboard->lpCoreKeyboard);
-      return FALSE;
-    }
+  DeleteCoreEnvironment(core_environment);
 
-    ResetCapsLock();
-    err_status = km_kbp_keyboard_get_imx_list(_td->lpActiveKeyboard->lpCoreKeyboard, &_td->lpActiveKeyboard->lpIMXList);
-    if (err_status != KM_KBP_STATUS_OK) {
-      SendDebugMessageFormat(0, sdmLoad, 0, "Keyman_ForceKeyboard Core: km_kbp_keyboard_get_imx_list failed with error status [%d]", err_status);
-      // Dispose of the keyboard to leave us in a consistent state
-      ReleaseKeyboardMemoryCore(&_td->lpActiveKeyboard->lpCoreKeyboard);
-      return FALSE;
-    }
+  if (err_status != KM_KBP_STATUS_OK) {
+    SendDebugMessageFormat(
+        0, sdmGlobal, 0, "Keyman_ForceKeyboard Core: km_kbp_state_create failed with error status [%d]", err_status);
+    // Dispose of the keyboard to leave us in a consitent state
+    ReleaseKeyboardMemoryCore(&_td->lpActiveKeyboard->lpCoreKeyboard);
+    goto fail;
+  }
 
-    LoadDLLsCore(_td->lpActiveKeyboard);
-    ActivateDLLs(_td->lpActiveKeyboard);
-    LoadKeyboardOptionsREGCore(_td->lpActiveKeyboard, _td->lpActiveKeyboard->lpCoreKeyboardState);
-    RefreshPreservedKeys(TRUE);
-    return TRUE;
-  } else {
-    if (LoadKeyboard(_td->ForceFileName, &_td->lpActiveKeyboard->Keyboard)) {
-      SendDebugMessageFormat(0, sdmGlobal, 0, "Keyman_ForceKeyboard: %s OK", FileName);
-      ResetCapsLock();
-      LoadDLLs(_td->lpActiveKeyboard);
-      ActivateDLLs(_td->lpActiveKeyboard);
-      LoadKeyboardOptions(_td->lpActiveKeyboard);  // I2437 - Crash unloading keyboard due to keyboard options not set
-      RefreshPreservedKeys(TRUE);
-      return TRUE;
-    }
+  ResetCapsLock();
+  err_status = km_kbp_keyboard_get_imx_list(_td->lpActiveKeyboard->lpCoreKeyboard, &_td->lpActiveKeyboard->lpIMXList);
+  if (err_status != KM_KBP_STATUS_OK) {
+    SendDebugMessageFormat(0, sdmLoad, 0, "Keyman_ForceKeyboard Core: km_kbp_keyboard_get_imx_list failed with error status [%d]", err_status);
+    // Dispose of the keyboard to leave us in a consistent state
+    ReleaseStateMemoryCore(&_td->lpActiveKeyboard->lpCoreKeyboardState);
+    ReleaseKeyboardMemoryCore(&_td->lpActiveKeyboard->lpCoreKeyboard);
+    goto fail;
   }
-	SendDebugMessageFormat(0,sdmGlobal,0,"Keyman_ForceKeyboard: %s FAIL", FileName);
+
+  LoadDLLs(_td->lpActiveKeyboard);
+  ActivateDLLs(_td->lpActiveKeyboard);
+  LoadKeyboardOptionsREGCore(_td->lpActiveKeyboard, _td->lpActiveKeyboard->lpCoreKeyboardState);
+  RefreshPreservedKeys(TRUE);
+  return TRUE;
+    // happy to use while(!done) pattern
+fail:
```

Comment:
I realised there was previously a possible memory leak here in not deleting `_td_>lpActiveKeyboard` I am using a goto here. If we don't like that I could maybe use while(!done) {} instead